pytorch_lightning/trainer/training_loop.py

```diff
@@ -558,21 +560,23 @@ def run_training_epoch(self):
         )
 
         should_check_val = self.should_check_val_fx(batch_idx, is_last_batch, on_epoch=True)
-        if should_check_val:
-            self.trainer.run_evaluation(on_epoch=True)
-
-            # reset stage to train
-            self.trainer._running_stage = RunningStage.TRAINING
-
         should_skip_eval = self.trainer.evaluation_loop.should_skip_evaluation(self.trainer.num_val_batches)
         should_train_only = self.trainer.disable_validation or should_skip_eval
 
-        if should_train_only:
-            # update epoch level lr_schedulers
+        # update epoch level lr_schedulers if no val loop outside train loop is triggered
+        if (val_loop_called and not should_check_val) or should_train_only:
             self.trainer.optimizer_connector.update_learning_rates(interval='epoch')
+
+        if should_train_only:
             self.check_checkpoint_callback(True)
             self.check_early_stopping_callback(True)
 
+        if should_check_val:
+            self.trainer.run_evaluation(on_epoch=True)
```

Comment:
are these checks guaranteed to be mutually exclusive? i'm updating to 1.2.4 and see a test failure with one of my modules where it looks like there's a gap: 
before: `run_evaluation` ran first, allowing the module to run the validation loop. inside validation step/validation epoch end, the module could log metrics. afterward, the checkpoint is force-run. so if the checkpoint configured a metric that was available only during validation, then somehow this still worked. 
by moving the run_evaluation to happen after the force checkpoint saving, we fail when looking up the monitor value

i think the correct fix is dropping the training loop force running checkpoint/early stopping callbacks here. they should be part of the callback.

Reply:
I believe if validation is disabled or val batches=0 then only we run the force checkpoint else it will call run_evaluation and the the callbacks will be called inside and monitor will be taken care of as expected. Can you paste a small example with your case where it doesn't work?

Reply:
It's not, see https://github.com/PyTorchLightning/pytorch-lightning/pull/7207 

this PR introduced the change where we call the evaluation loop after we call the checkpoint/early stopping callbacks from training. 

As a result, this check for `should_train_only` is incomplete - it inherently depends on the evaluation loop to populate `num_val_batches` correctly. in `run_evaluation` we set the validation dataloader, but this is too late as the validation dataloader is what's used to determine `should_skip_eval` above